File: Advanced/Section/CustomWordExport/C#/HtmlToOpenXmlExport/HtmlToOpenXmlExport.cs
using System;
using System.IO;
using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using GrapeCity.ActiveReports.Document;
using GrapeCity.ActiveReports.Export;
using GrapeCity.ActiveReports.Export.Html;
using GrapeCity.ActiveReports.Export.Html.Section;
using HtmlToOpenXml;

namespace GrapeCity.ActiveReports.Samples.WordExport
{
	public class HtmlToOpenXmlExport : IDocumentExport
	{
		private const string _htmlPath = @"..\..\htmlTemp\";
		private const string _htmlFilePath = _htmlPath + "temp.html";

		private void HtmlToOpenXml(string html, MemoryStream generatedDocument)
		{
			using (var buffer = ResourceHelper.GetStream("Resources.template.docx"))
				buffer.CopyToAsync(generatedDocument);

			generatedDocument.Position = 0L;
			using (var package = WordprocessingDocument.Open(generatedDocument, true))
			{
				MainDocumentPart mainPart = package.MainDocumentPart;
				if (mainPart == null)
				{
					mainPart = package.AddMainDocumentPart();
					new DocumentFormat.OpenXml.Wordprocessing.Document(new Body()).Save(mainPart);
				}

				var converter = new HtmlConverter(mainPart);
				converter.ImageProcessing = ImageProcessing.ManualProvisioning;
				converter.ProvisionImage += OnProvisionImage;

				var paragraphs = converter.Parse(html);

				Body body = mainPart.Document.Body;
				SectionProperties sectionProperties = GetLastChild<SectionProperties>(mainPart.Document.Body);

				for (int i = 0; i < paragraphs.Count; i++)
					body.Append(paragraphs[i]);

				if (sectionProperties != null)
				{
					sectionProperties.Remove();
					body.Append(sectionProperties);
				}

				mainPart.Document.Save();
			}
		}

		private T GetLastChild<T>(OpenXmlElement element) where T : OpenXmlElement
		{
			if (element == null) return null;

			for (int i = element.ChildElements.Count - 1; i >= 0; i--)
			{
				if (element.ChildElements[i] is T)
					return element.ChildElements[i] as T;
			}

			return null;
		}

		private void WriteAllBytes(Stream stream, MemoryStream generatedDocument)
		{
			var bytes = generatedDocument.ToArray();
			stream.Write(bytes, 0, bytes.Length);
		}

		private void WriteAllBytes(string destFilePath, MemoryStream generatedDocument)
		{
			File.WriteAllBytes(destFilePath, generatedDocument.ToArray());
		}

		private string HtmlExport(SectionDocument document, string pageRange)
		{
			if (!Directory.Exists(_htmlPath))
			{
				Directory.CreateDirectory(_htmlPath);
			}

			using (var htmlExport = new HtmlExport())
				htmlExport.Export(document, _htmlFilePath, pageRange);
			return GetHtmlString();
		}

		private string GetHtmlString()
		{
			using (var stream = new StreamReader(_htmlFilePath))
			{
				return stream.ReadToEnd();
			}
		}

		private void OnProvisionImage(object sender, ProvisionImageEventArgs e)
		{
			string filename = Path.GetFileName(e.ImageUrl.OriginalString);
			var imagePath = Path.Combine(_htmlPath, filename);
			if (!File.Exists(imagePath))
			{
				//e.Cancel = true;
				return;
			}

			e.Provision(File.ReadAllBytes(imagePath));
		}

		private void ClearHtmlPath()
		{
			foreach (var file in new DirectoryInfo(_htmlPath).GetFiles())
			{
				file.Delete();
			}
			Directory.Delete(_htmlPath);
		}

		void IDocumentExport.Export(SectionDocument document, string filePath, string pageRange)
		{
			var html = HtmlExport(document, pageRange);
			using (var generatedDocument = new MemoryStream())
			{
				HtmlToOpenXml(html, generatedDocument);
				WriteAllBytes(filePath, generatedDocument);
			}
			ClearHtmlPath();
		}

		void IDocumentExport.Export(SectionDocument document, string filePath)
		{
			((IDocumentExport)this).Export(document, filePath, string.Empty);
		}

		void IDocumentExport.Export(SectionDocument document, Stream outputStream)
		{
			((IDocumentExport)this).Export(document, outputStream, string.Empty);
		}

		void IDocumentExport.Export(SectionDocument document, Stream outputStream, string pageRange)
		{
			var html = HtmlExport(document, pageRange);
			using (var generatedDocument = new MemoryStream())
			{
				HtmlToOpenXml(html, generatedDocument);
				WriteAllBytes(outputStream, generatedDocument);
			}
			ClearHtmlPath();
		}

		void IDocumentExport.Export(SectionDocument document, IOutputHtml outputHandler, string pageRange)
		{
			throw new NotSupportedException("It's not allowed to use IOutputHtml");
		}
	}
}

File: Advanced/Section/CustomWordExport/VB.NET/HtmlToOpenXmlExport/HtmlToOpenXmlExport.vb
Imports GrapeCity.ActiveReports.Export.Html.Section
Imports GrapeCity.ActiveReports.Export.Html
Imports GrapeCity.ActiveReports.Export
Imports GrapeCity.ActiveReports.Document
Imports DocumentFormat.OpenXml.Wordprocessing
Imports DocumentFormat.OpenXml.Packaging
Imports System.IO
Imports DocumentFormat.OpenXml
Imports HtmlToOpenXml

Public Class HtmlToOpenXmlExport
	Implements IDocumentExport
	Private Const _htmlPath As String = "..\..\htmlTemp\"
	Private Const _htmlFilePath As String = _htmlPath + "temp.html"

	Private Sub HtmlToOpenXml(html As String, generatedDocument As MemoryStream)
		Using buffer = ResourceHelper.GetStream("template.docx")
			buffer.CopyToAsync(generatedDocument)
		End Using

		generatedDocument.Position = 0L
		Using package = WordprocessingDocument.Open(generatedDocument, True)
			Dim mainPart As MainDocumentPart = package.MainDocumentPart
			If mainPart Is Nothing Then
				mainPart = package.AddMainDocumentPart()
				Dim doc = New Wordprocessing.Document(New Body())
				doc.Save(mainPart)
			End If

			Dim converter As New HtmlConverter(mainPart)
			converter.ImageProcessing = ImageProcessing.ManualProvisioning
			AddHandler converter.ProvisionImage, AddressOf OnProvisionImage

			Dim paragraphs As IList(Of OpenXmlCompositeElement) = converter.Parse(html)

			Dim body As Body = mainPart.Document.Body
			Dim sectionProperties As SectionProperties = GetLastChild(Of SectionProperties)(mainPart.Document.Body)

			Dim i As Integer = 0
			While i < paragraphs.Count
				body.Append(paragraphs(i))
				i += 1
			End While

			If sectionProperties IsNot Nothing Then
				sectionProperties.Remove()
				body.Append(sectionProperties)
			End If

			mainPart.Document.Save()
		End Using
	End Sub

	Private Function GetLastChild(Of T As OpenXmlElement)(element As OpenXmlElement) As T
		If element Is Nothing Then
			Return Nothing
		End If

		Dim i As Integer = element.ChildElements.Count - 1
		While i >= 0
			If TypeOf element.ChildElements(i) Is T Then
				Return TryCast(element.ChildElements(i), T)
			End If
			i -= 1
		End While

		Return Nothing
	End Function

	Private Sub WriteAllBytes(stream As Stream, generatedDocument As MemoryStream)
		Dim bytes = generatedDocument.ToArray()
		stream.Write(bytes, 0, bytes.Length)
	End Sub

	Private Sub WriteAllBytes(destFilePath As String, generatedDocument As MemoryStream)
		File.WriteAllBytes(destFilePath, generatedDocument.ToArray())
	End Sub

	Private Function HtmlExport(document As SectionDocument, pageRange As String) As String
		If Not Directory.Exists(_htmlPath) Then
			Directory.CreateDirectory(_htmlPath)
		End If
		Using export = New HtmlExport
			export.Export(document, _htmlFilePath, pageRange)
		End Using
		Return GetHtmlString()
	End Function

	Private Function GetHtmlString() As String
		Using stream As New StreamReader(_htmlFilePath)
			Return stream.ReadToEnd()
		End Using
	End Function

	Private Sub OnProvisionImage(sender As Object, e As ProvisionImageEventArgs)
		Dim filename As String = Path.GetFileName(e.ImageUrl.OriginalString)
		Dim imagePath = Path.Combine(_htmlPath, filename)
		If Not File.Exists(imagePath) Then
			'e.Cancel = True
			Return
		End If

		e.Provision(File.ReadAllBytes(imagePath))
	End Sub

	Private Sub ClearHtmlPath()
		For Each file In New DirectoryInfo(_htmlPath).GetFiles()
			file.Delete()
		Next
		Directory.Delete(_htmlPath)
	End Sub

	Sub Export(document As SectionDocument, filePath As String, pageRange As String) Implements IDocumentExport.Export
		Dim html = HtmlExport(document, pageRange)
		Using generatedDocument As New MemoryStream()
			HtmlToOpenXml(html, generatedDocument)
			WriteAllBytes(filePath, generatedDocument)
		End Using
		ClearHtmlPath()
	End Sub

	Sub Export(document As SectionDocument, filePath As String) Implements IDocumentExport.Export
		CType(Me, IDocumentExport).Export(document, filePath, String.Empty)
	End Sub

	Sub Export(document As SectionDocument, outputStream As Stream) Implements IDocumentExport.Export
		CType(Me, IDocumentExport).Export(document, outputStream, String.Empty)
	End Sub

	Sub Export(document As SectionDocument, outputStream As Stream, pageRange As String) Implements IDocumentExport.Export
		Dim html = HtmlExport(document, pageRange)
		Using generatedDocument As New MemoryStream()
			HtmlToOpenXml(html, generatedDocument)
			WriteAllBytes(outputStream, generatedDocument)
		End Using
		ClearHtmlPath()
	End Sub

	Sub Export(document As SectionDocument, outputHandler As IOutputHtml, pageRange As String) Implements IDocumentExport.Export
		Throw New NotSupportedException("It's not allowed to use IOutputHtml")
	End Sub
End Class
